src/parser/scanner.lex

```diff
@@ -0,0 +1,179 @@
+                                    }
+                                    if (yyleng - i > 22 ||
+                                            (yyleng - i == 22 && yytext[i] != '1')) {
+                                        throw ExprParser::syntax_error(*yylloc, "Out of range:");
+                                    }
+                                }
+                                uint64_t val = 0;
+                                sscanf(yytext, "%lo", &val);
+                                if (val > MAX_ABS_INTEGER) {
+                                    throw ExprParser::syntax_error(*yylloc, "Out of range:");
+                                }
+                                yylval->intval = static_cast<int64_t>(val);
+                                return TokenType::INTEGER;
+                            }
+{DEC}+                      {
+                                try {
+                                    folly::StringPiece text(yytext, yyleng);
+                                    uint64_t val = folly::to<uint64_t>(text);
+                                    if (val > MAX_ABS_INTEGER) {
+                                        throw ExprParser::syntax_error(*yylloc, "Out of range:");
+                                    }
+                                    yylval->intval = val;
+                                } catch (...) {
+                                    throw ExprParser::syntax_error(*yylloc, "Out of range:");
+                                }
+                                return TokenType::INTEGER;
+                            }
+{DEC}+\.{DEC}*              {
+                                try {
+                                    folly::StringPiece text(yytext, yyleng);
+                                    yylval->doubleval = folly::to<double>(text);
+                                } catch (...) {
+                                    throw ExprParser::syntax_error(*yylloc, "Out of range:");
+                                }
+                                return TokenType::DOUBLE;
+                            }
+{DEC}*\.{DEC}+              {
+                                try {
+                                    folly::StringPiece text(yytext, yyleng);
+                                    yylval->doubleval = folly::to<double>(text);
+                                } catch (...) {
+                                    throw ExprParser::syntax_error(*yylloc, "Out of range:");
+                                }
+                                return TokenType::DOUBLE;
+                            }
+
+\${STRING}                  { yylval->strval = new std::string(yytext + 1, yyleng - 1); return TokenType::VARIABLE; }
+
+[ \r\t]                     { }
+\n                          {
+                                yylineno++;
+                                yylloc->lines(yyleng);
+                            }
+
+%%
+
+namespace nebula_chaos {
+
+ExprScanner::ExprScanner(std::istream* in,
```

Comment:
Do you remember what's the difference between define `ExprScanner` ctor at the end of `.lex`? I mean we didn't define it in `nebula` repo.